posts/inside-rust/2020-10-30-exploring-pgo-for-the-rust-compiler.md

```diff
@@ -0,0 +1,299 @@
+    Just run the compiler as you always would (e.g. via Cargo) and the profile data will show up in the `<PROFDATA_DIR>` we specified in the `-fprofile-generate` flag above.
+    In order to make the collected data as useful as possible, we should try to exercise all the common code paths within the compiler.
+    I typically use the "standard" [rustc-perf][rustc-perf] benchmark suite for this purpose, which includes debug builds, optimized builds, check builds, both incremental and non-incremental.
+    After this is done, you will find a number of `.profraw` files in `<PROFDATA_DIR>`.
+    [As described in the Clang user manual][llvm-profdata] these `.profraw` files need to be merged into a single `.profdata` file by using the `llvm-profdata` tool that comes with your Clang installation:
+
+    ```bash
+    $ cd <PROFDATA_DIR>
+    $ llvm-profdata merge -output=rustc-llvm.profdata *.profraw
+    ```
+
+[rustc-perf]: https://github.com/rust-lang/rustc-perf
+[llvm-profdata]: https://clang.llvm.org/docs/UsersManual.html#cmdoption-fprofile-generate
+
+3. Now that the combined profile data from all *rustc* invocations can be found in `<PROFDATA_DIR>/rustc-llvm.profdata` it is time to re-compile LLVM and *rustc* again, this time instructing Clang to make use of this valuable new information.
+    To this end, we modify `config.toml` as follows:
+
+    ```toml
+    [llvm]
+    # Instead of -fprofile-generate, we now pass -fprofile-use to Clang
+    cflags = "-fprofile-use=<PROFDATA_DIR>/rustc-llvm.profdata"
+    cxxflags = "-fprofile-use=<PROFDATA_DIR>/rustc-llvm.profdata"
+    ```
+
+    Now we make sure that LLVM is properly rebuilt by deleting the old version and build everything again:
+
+    ```bash
+    $ cd $RUST_PROJECT_ROOT
+    $ rm -rf ./build/x86_64-unknown-linux-gnu/llvm
+    $ ./x.py build
+    ```
+
+    Once this is done, we have a Rust compiler with PGO-optimized LLVM. Congratulations!
+
+
+#### PGO-optimized LLVM -- Benchmark Results
+
+As mentioned above Firefox build times have improved by up to 9% with a PGOed compiler.
+Clang's own documentation even [reports][clang-pgo-20] an up to 20% improvement.
+The best way we have for assessing the Rust compiler's performance is the [rustc-perf][rustc-perf] benchmark suite.
+Since compiling with PGO does not quite fit with how the Rust project's CI works, we cannot use the [perf.rust-lang.org][perf.rlo] version of the benchmark suite.
+Fortunately, thanks to [good documentation][rustc-perf-local], running the benchmarks locally is straightforward enough.
+Here's a glance at the effect that a PGOed LLVM has on *rustc*'s performance:
+
+[![Performance improvements gained from apply PGO to LLVM][rustc-perf-pgo-llvm-thumb]][rustc-perf-pgo-llvm]
+[rustc-perf-pgo-llvm-thumb]: /images/inside-rust/2020-10-30-exploring-pgo-for-the-rust-compiler/rustc-perf-pgo-llvm-thumb.png
+[rustc-perf-pgo-llvm]: /images/inside-rust/2020-10-30-exploring-pgo-for-the-rust-compiler/rustc-perf-pgo-llvm.png
```

Comment:
I could send you the `results.db`. Would that help?

Reply:
Yeah, it shouldn't be too hard to pull results from there and into the postgres db used by the official backend, which would mean that we can provide URLs here rather than pulling images (which should be all around nicer I imagine).

I expect to name the artifacts e.g. "pgo-llvm" as you've done here fwiw.